### ISSUE TITLE
Nested calls to synchronize cause exponential delays (retries**2)

### ISSUE THREAD

**Issue by TylerRick** [teamcapybara/capybara#725]
### 128-second hangs before finally failing with `ElementNotFound`

I've noticed that sometimes capybara will **appear to hang forever** (it's not actually forever but it's a really long time) when it looks for an element and can't find it. This is pretty frustrating when my cucumber test run just hangs indefinitely for no apparent reason.

A [simple test scenario](https://github.com/TylerRick/capybara_nested_synchronize_bug/blob/master/features/step_definitions/reproduce_bug_steps.rb) that _should_ only take about [`11 s`](https://gist.github.com/2893016) to fail ends up taking a whopping [`128 s`](https://gist.github.com/2892966) before it finally fails with a `Capybara::ElementNotFound` error !

(In my case, I'm actually doing something where I _expect_ it to not find the element, so I have to wait that long before I can rescue the `Capybara::ElementNotFound` error and move on with the rest of my test...)

After hours of debugging, I managed to distill this down to a [reproducible scenario](https://github.com/TylerRick/capybara_nested_synchronize_bug). But I still don't understand completely what's going on...

It seems to be the combination of `within` and a `find` that it doesn't like... although I've done  `find`s inside of `within`s  without problems before so I don't know what exactly is wrong with this scenario...
## So what's going on?

From what I can tell, it looks like this code in synchronize that is supposed to retry a search query a certain number of times (40), ends up getting called somewhat recursively and so it ends up retrying the query 40*40 times:

``` ruby
      def synchronize(seconds=Capybara.default_wait_time)
        retries = (seconds.to_f / 0.05).round

        begin
          yield
        rescue => e
          raise e unless driver.wait?
          raise e unless driver.invalid_element_errors.include?(e.class) or e.is_a?(Capybara::ElementNotFound)
          raise e if retries.zero?
          sleep(0.05)
          reload if Capybara.automatic_reload
          retries -= 1
          retry
        end
      end
```

Instead of just doing `retries` number of retries (`Capybara.default_wait_time / 0.05` == `40` in my case) , it appears to be doing approximately `retries**2` (`40**2` == `1600`) number of retries!
## Debug output

You can observe this in the output from running rake cucumber with some debug output added (see [workaround.rb](https://github.com/TylerRick/capybara_nested_synchronize_bug/blob/master/lib/workaround.rb) for the debugging code that created the output and [gist](https://gist.github.com/2892966) for the full output)...

The first section (before the "=========") shows how it is _supposed_ to work, and does normally work... You can see the retries number counting from 40 all the way down to 0.

The second section (after the "=========") shows what happens when it _doesn't_ work how I think it's supposed to work... You can see the retries number from the **inner loop** (lines starting with 1) counting from 40 all the way down to 0. But as soon as the inner loop completes, the **outer loop** (lines starting with 0) begin counting down, each time executing the inner loop all over again...

```
first number indicates the nesting level
\/
1 retries=1 e=#<Capybara::ElementNotFound: no select box with id, name, or label 'select_that_doesnt_exist' found>
1 retries=0 e=#<Capybara::ElementNotFound: no select box with id, name, or label 'select_that_doesnt_exist' found>
0 retries=40 e=#<Capybara::ElementNotFound: no select box with id, name, or label 'select_that_doesnt_exist' found>
1 retries=40 e=#<Capybara::ElementNotFound: no select box with id, name, or label 'select_that_doesnt_exist' found>
1 retries=39 e=#<Capybara::ElementNotFound: no select box with id, name, or label 'select_that_doesnt_exist' found>
1 retries=38 e=#<Capybara::ElementNotFound: no select box with id, name, or label 'select_that_doesnt_exist' found>
...
```

Here's a [gist](https://gist.github.com/2892996) showing the relevant lines from `caller` that show where synchronize is being called.
## Workaround

My [workaround](https://github.com/TylerRick/capybara_nested_synchronize_bug/blob/master/lib/workaround.rb) basically involved preventing the **inner** loops from looping at all if it detected we were already in an **outer** synchronize retry "loop"

Basically I just changed:

``` ruby
          raise e if retries.zero?
```

to:

``` ruby
          raise e if retries.zero? or @synchronize_nesting > 0
```
## Solution?

My workaround probably isn't the correct solution to the problem, since I don't really understand what's going on but it seems to at least fix the problem for me...

Can you please provide/suggest a better solution or tell me what I'm doing wrong?

(Note: This only affects the Selenium Webdriver driver, of course, not Rack Test.)
